### PR TITLE
Make threshold configurable for TreeOverlay

### PR DESCRIPTION
**Pull request by youngjoon-lee**
Based on PR #425 

Like FlatOverlay, we should use 100% leader threshold for TreeOverlay as well (for testing) because we don't have the block catch-up mechanism yet.

### CODE REVIEW

File: consensus-engine/src/overlay/threshold.rs [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    Fraction::new(
+        LEADER_SUPER_MAJORITY_THRESHOLD_NUM,
+        LEADER_SUPER_MAJORITY_THRESHOLD_DEN,
+    )
+}
+
+pub fn apply_threshold(size: usize, threshold: GenericFraction<u64>) -> usize {
+    // `threshold` is a tuple of (num, den) where `num/den` is the super majority threshold
+    (Fraction::from(size) * threshold)
+        .ceil()
```
Comment (youngjoon-lee) [inline consensus-engine/src/overlay/threshold.rs:16]: Previously, FlatOverlay was using `floor()`, and TreeOverlay was using `x * 2/3 + 1`. I thought it's better to have a common function like this and chose to use `ceil()` to guarantee that the threshold is never smaller than the super majority.

**Review by bacv — APPROVED** [review]

**Review by al8n — APPROVED** [review]

File: consensus-engine/src/overlay/tree_overlay/overlay.rs [hunk 2]
```diff
@@ -158,8 +171,7 @@ where
         // });
         // let root_size = self.root_committee().len();
         // let committee_size = root_size + children_size;
-        let committee_size = self.root_committee().len();
-        (committee_size * 2 / 3) + 1
+        apply_threshold(self.root_committee().len(), self.leader_threshold)
```
Comment (Zeegomo) [inline consensus-engine/src/overlay/tree_overlay/overlay.rs:174]: I think it's sensible that the threshold is configurable for both the leader supermajority and the 'standard' supermajority. For the flat overlay I only added that for the leader supermajority as the other one is always 0.
Here, I would suggest we use a single threshold and apply it to both

**Review by Zeegomo — CHANGES_REQUESTED** [review]

File: consensus-engine/src/overlay/tree_overlay/overlay.rs (hunk 2 above)
Comment (youngjoon-lee) [reply consensus-engine/src/overlay/tree_overlay/overlay.rs]: Good point. Thank you!: f1e3524

**Review by danielSanchezQ — APPROVED** [review]
Nice, thanks!

**Review by Zeegomo — APPROVED** [review]

**Comment by youngjoon-lee** [comment]
I'm not merging this PR now because CI fails, and I'm not sure if this failure is caused by this PR or not.
```
waiting... 1 | 1 | 1 | 1 | 1 | 1 | 1 | 1 | 0
waiting... 1 | 1 | 1 | 1 | 1 | 1 | 1 | 1 | 0
waiting... 1 | 1 | 1 | 1 | 1 | 1 | 1 | 1 | 0
waiting... 1 | 1 | 1 | 1 | 1 | 1 | 1 | 1 | 0
thread 'ten_nodes_one_down' panicked at 'timed out waiting for nodes to reach view 20', tests/src/tests/unhappy.rs:25:24
```

**Comment by youngjoon-lee** [comment]
> I'm not merging this PR now because CI fails, and I'm not sure if this failure is caused by this PR or not. [Test Suite (waku, ubuntu-latest)](https://github.com/logos-co/nomos-node/actions/runs/6297665462/job/17097238823?pr=426#logs)
> 
> ```
> waiting... 1 | 1 | 1 | 1 | 1 | 1 | 1 | 1 | 0
> waiting... 1 | 1 | 1 | 1 | 1 | 1 | 1 | 1 | 0
> waiting... 1 | 1 | 1 | 1 | 1 | 1 | 1 | 1 | 0
> waiting... 1 | 1 | 1 | 1 | 1 | 1 | 1 | 1 | 0
> thread 'ten_nodes_one_down' panicked at 'timed out waiting for nodes to reach view 20', tests/src/tests/unhappy.rs:25:24
> ```
> 
> This isn't even reproduced in my local.

I've also found that the same failure sometimes occur in the `ten_nodes_happy` test. I'm debugging it. 

One interesting thing is that the failure is not reproduced in my local (A Macbook with 10 cores), but it's reproduced in my Ubuntu VM (with 2 cores) in my Macbook, which has the similar resources as Github Actions hosts.

**Comment by youngjoon-lee** [comment]
> I've also found that the same failure sometimes occur in the `ten_nodes_happy` test. I'm debugging it.
> 
> One interesting thing is that the failure is not reproduced in my local (A Macbook with 10 cores), but it's reproduced in my Ubuntu VM (with 2 cores) in my Macbook, which has the similar resources as Github Actions hosts.

I'm debugging this at https://github.com/logos-co/nomos-node/issues/440. From my understanding so far, this failure is not highly related to this PR. But, I'm not sure yet.

**Pull request merged**